File: src/main/java/raccoonlang/parser/ast/DataClassDefinitionAstNode.java
package raccoonlang.parser.ast;

import raccoonlang.parser.Parser;
import raccoonlang.tokenizer.Token;
import raccoonlang.tokenizer.TokenType;

public class DataClassDefinitionAstNode {

    public ModifiersAstNode modifiersAstNode;
    public Token name;
    public GenericTypesAstNode genericTypesAstNode;
    public FunctionParametersAstNode functionParametersAstNode;
    public GenericTypeConstraintsAstNode genericTypeConstraintsAstNode;
    public DataClassBodyAstNode dataClassBodyAstNode;

    public static DataClassDefinitionAstNode tryParse(Parser parser) {
        Parser.ParserState parserState = parser.shelfState();
        DataClassDefinitionAstNode dataClassDefinitionAstNode = new DataClassDefinitionAstNode();

        dataClassDefinitionAstNode.modifiersAstNode = ModifiersAstNode.parse(parser);

        if(parser.peek().getTokenType() != TokenType.DATA) {
            parser.restore(parserState);
            return null;
        }
        parser.take();
        parser.take(TokenType.CLASS);

        dataClassDefinitionAstNode.name = parser.take(TokenType.IDENTIFIER);
        dataClassDefinitionAstNode.genericTypesAstNode = GenericTypesAstNode.tryParse(parser);
        dataClassDefinitionAstNode.functionParametersAstNode = FunctionParametersAstNode.parse(parser);
        dataClassDefinitionAstNode.genericTypeConstraintsAstNode = GenericTypeConstraintsAstNode.parse(parser);
        dataClassDefinitionAstNode.dataClassBodyAstNode = DataClassBodyAstNode.parse(parser);

        return dataClassDefinitionAstNode;
    }

    @Override
    public String toString() {
        return "DataClassDefinitionAstNode{" +
                "modifiersAstNode=" + modifiersAstNode +
                ", name=" + name +
                ", genericTypesAstNode=" + (genericTypesAstNode == null ? "" : genericTypesAstNode.toString()) +
                ", functionParametersAstNode=" + functionParametersAstNode +
                ", genericTypeConstraintsAstNode=" + genericTypeConstraintsAstNode +
                ", dataClassBodyAstNode=" + dataClassBodyAstNode +
                '}';
    }
}

File: Parsing/AST/DataClassDefinitionAstNode.cs
using LLVMSharp;

namespace Raccoonlang.Parsing.AST;

using Tokenizing;

public class DataClassDefinitionAstNode : ITypeDef
{
    public ModifiersAstNode? Modifiers { get; set; }
    public Token? Name { get; set; }
    public GenericTypesAstNode? GenericTypes { get; set; }
    public FunctionParametersAstNode? Parameters { get; set; }
    public GenericTypeConstraintsAstNode? GenericConstraints { get; set; }
    public DataClassBodyAstNode? BodyContainer { get; set; }

    public static DataClassDefinitionAstNode? TryParse(Parser parser)
    {
        DataClassDefinitionAstNode node = new DataClassDefinitionAstNode();

        var parserState = parser.ShelfState();
        node.Modifiers = ModifiersAstNode.Parse(parser);

        if (parser.Peek().Type != TokenType.Data)
        {
            parser.RestoreState(parserState);
            return null;
        }
        parser.Take(TokenType.Data);
        parser.Take(TokenType.Class);
        
        node.Name = parser.Take(TokenType.Identifier);
        node.GenericTypes = GenericTypesAstNode.TryParse(parser);
        node.Parameters = FunctionParametersAstNode.Parse(parser);
        node.GenericConstraints = GenericTypeConstraintsAstNode.Parse(parser);
        node.BodyContainer = DataClassBodyAstNode.Parse(parser);

        return node;
    }

    public override string ToString()
    {
        return "DataClassDefinitionAstNode{" +
                "modifiersAstNode=" + Modifiers +
                ", name=" + Name +
                ", genericTypesAstNode=" + GenericTypes +
                ", functionParametersAstNode=" + Parameters +
                ", genericTypeConstraintsAstNode=" + GenericConstraints+
                ", dataClassBodyAstNode=" + BodyContainer +
                "}";
    }

    public void Compile(LLVMContextRef context, IRBuilder irBuilder, Module module)
    {
        throw new NotImplementedException();
    }
}